File: app/src/main/java/com/anode/anode/MainActivity.java
package com.anode.anode;

import android.content.Intent;
import android.content.SharedPreferences;
import android.content.res.AssetManager;
import android.net.VpnService;
import android.os.Bundle;

import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.android.material.snackbar.Snackbar;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;

import android.view.View;
import android.view.Menu;
import android.view.MenuItem;
import android.util.Log;
import android.widget.Button;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

public class MainActivity extends AppCompatActivity {

    private static String LOGTAG = "MainActivity";
    private FloatingActionButton button_start_test;
    private Button button_get_peers;
    SharedPreferences prefs = null;
    AnodeUtil util;

    protected void InitializeApp(){
        //Create files folder
        getApplication().getFilesDir().mkdir();
        //Copy cjdroute
        InputStream in = null;
        OutputStream out = null;
        try {
            AssetManager am = getBaseContext().getAssets();
            //Read architecture
            String arch = System.getProperty("os.arch");
            Log.i(LOGTAG,"OS Architecture: "+arch);
            if (arch.contains("x86") || arch.contains("i686"))
            {
                in = am.open("x86/cjdroute");
            } else if (arch.contains("arm64-v8a")){
                in = am.open("arm64-v8a/cjdroute");
            } else if (arch.contains("armeabi")) {
                in = am.open("armeabi-v7a/cjdroute");
            } else {
                //Unknown architecture
                Log.i(LOGTAG,"Unknown CPU architecture");
                return;
            }

            out = new FileOutputStream(getApplication().getFilesDir()+"/cjdroute");
            byte[] buffer = new byte[1024];
            int read;
            while ((read = in.read(buffer)) != -1 ){
                out.write(buffer, 0, read);
            }
            in.close();
            out.close();
            //Set permissions
            File file = new File(getApplication().getFilesDir()+"/cjdroute");
            file.setExecutable(true);
        } catch (IOException e) {
            e.printStackTrace();
            Log.e(LOGTAG, "Failed to copy cjdroute file", e);
        }

        //Create conf
        util.Genconf();

        //Modify default conf file
        //cjdns.ModifyConfFile();
        util.ModifyJSONConfFile();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        util = new AnodeUtil();
        prefs = getSharedPreferences("com.anode.anode", MODE_PRIVATE);
        if (prefs.getBoolean("firstrun", true)) {
            InitializeApp();
            prefs.edit().putBoolean("firstrun", false).commit();
        }

        button_start_test = findViewById(R.id.button_start_test);
        button_get_peers = findViewById(R.id.button_getPeers);
        //DEBUG
        //cjdns.ModifyJSONConfFile();
        button_start_test.setOnClickListener(new View.OnClickListener() {

            @Override
            public void onClick(View view) {
                Snackbar.make(view, "Launching cjdroute...", Snackbar.LENGTH_LONG)
                        .setAction("Action", null).show();
                Intent intent = VpnService.prepare(getApplicationContext());
                if (intent != null) {
                    startActivityForResult(intent, 0);
                } else {
                    onActivityResult(0, RESULT_OK, null);
                }
            }
        });

    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_settings) {
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (resultCode == RESULT_OK) {
            Intent intent = new Intent(this, AnodeVpnService.class);
            startService(intent);
        }
    }

    public void showMessageSnackbar(String msg)
    {
        View view = findViewById(button_start_test.getId());
        Snackbar.make(view, msg, Snackbar.LENGTH_LONG)
                .setAction("Action", null).show();
    }

}

File: app/src/main/java/com/anode/anode/MainActivity.kt
package com.anode.anode

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.text.Html
import android.text.Spanned
import android.text.method.LinkMovementMethod
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.text.HtmlCompat
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.google.android.material.snackbar.Snackbar
import org.w3c.dom.Text
import java.io.*


class MainActivity : AppCompatActivity() {

    private var util: AnodeUtil? = null

    private fun initializeApp() {
        //Create files folder
        application.filesDir.mkdir()
        val cjdrouteFile = File(AnodeUtil().CJDNS_PATH+"/"+AnodeUtil().cjdrouteConfFile)
        if (!cjdrouteFile.exists()) {
            //Copy cjdroute
            var `in`: InputStream?
            var out: OutputStream?
            try {
                val am = baseContext.assets
                //Read architecture
                val arch = System.getProperty("os.arch")
                Log.i(LOGTAG, "OS Architecture: $arch")
                `in` = if (arch!!.contains("x86") || arch.contains("i686")) {
                    am.open("i686/16/cjdroute")
                } else if (arch.contains("arm64-v8a") || arch.contains("aarch64")) {
                    am.open("aarch64/21/cjdroute")
                } else if (arch.contains("armeabi") || arch.contains("armv7a")) {
                    am.open("armv7a/16/cjdroute")
                } else if (arch.contains("x86_64")) {
                    am.open("X86_64/21/cjdroute")
                } else { //Unknown architecture
                    Log.i(LOGTAG, "Incompatible CPU architecture")
                    return
                }
                out = FileOutputStream(application.filesDir.toString() + "/cjdroute")
                val buffer = ByteArray(1024)
                var read: Int
                while (`in`.read(buffer).also { read = it } != -1) {
                    out.write(buffer, 0, read)
                }
                `in`.close()
                out.close()
                //Set permissions
                val file = File(application.filesDir.toString() + "/cjdroute")
                file.setExecutable(true)
            } catch (e: IOException) {
                e.printStackTrace()
                Log.e(LOGTAG, "Failed to copy cjdroute file", e)
            }
        }
        //Create and initialize conf file
        if (!File(application.filesDir.toString()+"/"+AnodeUtil().cjdrouteConfFile).exists()) {
            util!!.initializeCjdrouteConfFile()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {

        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        util = AnodeUtil()
        val prefs = getSharedPreferences("com.anode.anode", Context.MODE_PRIVATE)

        //Initialize the app by copying cjdroute and generating the conf file
        initializeApp()

        /* We may need the first run check in the future... */
        /*
        if (prefs.getBoolean("firstrun", true)) {
            prefs.edit().putBoolean("firstrun", false).commit()
        } */
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean { // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean { // Handle action bar item clicks here. The action bar will
// automatically handle clicks on the Home/Up button, so long
// as you specify a parent activity in AndroidManifest.xml.
        val id = item.itemId
        return if (id == R.id.action_settings) {
            true
        } else super.onOptionsItemSelected(item)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (resultCode == Activity.RESULT_OK) {
            val intent = Intent(this, AnodeVpnService::class.java)
            startService(intent)
        }
    }

    companion object {
        private const val LOGTAG = "MainActivity"
    }
}